src/hotspot/share/gc/shenandoah/shenandoahFreeSet.hpp

```diff
@@ -220,17 +223,41 @@ class ShenandoahRegionPartitions {
 
   inline size_t available_in(ShenandoahFreeSetPartitionId which_partition) const {
     assert (which_partition < NumPartitions, "selected free set must be valid");
-    return _capacity[int(which_partition)] - _used[int(which_partition)];
+    shenandoah_assert_heaplocked();
+    assert(_available[int(which_partition)] == _capacity[int(which_partition)] - _used[int(which_partition)],
+           "Expect available (%zu) equals capacity (%zu) - used (%zu) for partition %s",
+           _available[int(which_partition)], _capacity[int(which_partition)], _used[int(which_partition)],
+           partition_membership_name(ssize_t(which_partition)));
+    return _available[int(which_partition)];
+  }
+
+  // Acquire heap lock and return available_in, assuming heap lock is not acquired by the caller.
+  inline size_t available_in_under_lock(ShenandoahFreeSetPartitionId which_partition) const {
```

Comment:
This name confuses me: `available_in_under_lock`. Should it be called `available_without_lock` or `available_no_lock`? Or, switch it with `available_in` (which asserts that the heap lock is held). I see that it takes the lock, but this is only to make the assertion.

Reply:
Thanks for review and comments.  I'll change the name.  It follows a pattern that is admittedly very confusing...

Reply:
changing to available_in_not_locked()